_posts/2024-01-23-ensuring-code-quality-a-guide-to-dynamic-git-hub-pull-request-gates.markdown

```diff
@@ -0,0 +1,309 @@
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      # https://github.com/marketplace/actions/changed-files
+      - uses: tj-actions/changed-files@v41.0.1
+        id: relevant-files-changed
+        with:
+          files: |
+            src/**
+            tests/**
+ 
+  build-deploy:
+    needs: check-if-relevant
+    if: needs.check-if-relevant.outputs.is-relevant == 'true'
+    runs-on: ubuntu-latest
+    steps:
+      [...]
+    
+  test:
+    needs: build-deploy
+    runs-on: ubuntu-latest
+    steps:
+      [...]
+```
+
+## Reusable Workflow as Status Check
+
+The approach works fine for standard workflows, where any job can be configured as status check.
+However, it doesn't work for [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows).
+
+In this example the test workflow utilizes reusable workflows for the build and deploy and the test steps.
+
+```yaml
+on:
+  pull_request:
+    
+jobs:
+  check-if-relevant:
+    runs-on: ubuntu-latest
+    outputs:
+      is-relevant: ${{ '{{' }} steps.relevant-files-changed.outputs.any_changed }}
+    steps:
+     [...]
+ 
+  build-deploy:
+    needs: check-if-relevant
+    if: needs.check-if-relevant.outputs.is-relevant == 'true'
+    uses: ./.github/workflows/release.build_deploy.yml
+    secrets: inherit    
+
+  test:
+    needs: build-deploy
+    uses: ./.github/workflows/test.execution.yml
+    secrets: inherit
+    with:
+      app-url: ${{ '{{' }} needs.build-deploy.outputs.app-url }}
+```
+
+The calling job `test` can't be set as status check because GitHub ignores its status.
+
+Configuring a job in the called reusable workflow `test.execution.yml` is not an option either.<br>
+If `test` in the calling workflow would be skipped, the pull request would wait forever for the status check in the 
+reusable workflow as the skipping happens on workflow parent level, not child level.
+
+The workaround is to configure an additional, standard job `test-completed` that concludes the workflow as status 
+check and fails if the `test` failed before.<br>
+Setting `if: always()` ensures that the job is always executed, even if the `test` job failed.
+
+```yaml
+on:
+  pull_request:
+    
+jobs:
+  check-if-relevant:
+    runs-on: ubuntu-latest
+    outputs:
+      is-relevant: ${{ '{{' }} steps.relevant-files-changed.outputs.any_changed }}
+    steps:
+     [...]
+ 
+  build-deploy:
+    needs: check-if-relevant
+    if: needs.check-if-relevant.outputs.is-relevant == 'true'
+    uses: ./.github/workflows/release.build_deploy.yml
+    secrets: inherit  
+
+  test:
+    needs: build-deploy
+    uses: ./.github/workflows/test.execution.yml
+    secrets: inherit
+    with:
+      app-url: ${{ '{{' }} needs.build-deploy.outputs.app-url }}
+
+  test-completed:
+    needs: test
+    runs-on: 'ubuntu-latest'
+    if: always()
+    steps:
+      - name: Check test job status
+        if: needs.test.result == 'failure'
+        run: exit 1
+```
```

Comment:
Maybe add a small section of what is the adv/disadv of this approach(I only see these for the following alternate approach(or does the adv/disadv apply to both?).

Reply:
I changed the titles of the two workflows and renamed the adv/disadv of the alternative approach to `Comparison`.